proposals/12513-pvc-artifact-storage/README.md

```diff
@@ -0,0 +1,1867 @@
+
+### Non-Goals
+
+1. **Replace object storage** - filesystem storage is an additional option, not a replacement
+2. **Cross-namespace artifact sharing** - Initial implementation focuses on namespace isolation
+3. **Performance optimization for large-scale workloads** - filesystem storage may not match object storage performance for massive datasets or high concurrency scenarios
+4. **Storage migration tools** - Migrating from S3 to PVC/filesystem is not supported
+5. **Advanced storage features** - No versioning, replication, or geo-distribution initially
+
+## Proposal
+
+### Overview
+
+This KEP proposes adding a new artifact storage backend that uses filesystem storage (primarily Kubernetes `PersistentVolumeClaims`) instead of object storage. The implementation will:
+
+1. Create one PVC per namespace for artifact storage
+2. Use configurable access mode with sensible defaults (RWO)
+3. Organize artifacts in a filesystem hierarchy within the PVC
+4. Provide transparent access through the existing KFP artifact APIs with new `kfp-artifacts://` URI scheme
+5. Maintain compatibility with existing pipeline definitions that don't have hardcoded storage paths
+6. Support separate scaling of artifact serving through artifacts-only instances
+7. Update the UI to seamlessly handle artifact downloads from filesystem storage
+
+### User Stories
+
+#### Story 1: User Running Pipelines on Local Kubernetes
+
+As a user with KFP on kind/minikube/k3s, I want my pipeline artifacts to automatically use the local cluster's default `StorageClass` via the `kfp-artifacts://` scheme, so that I can develop pipelines offline without any storage configuration.
+
+**Acceptance Criteria:**
+
+- KFP works out-of-the-box with filesystem storage on local clusters
+- Artifacts are stored using the `kfp-artifacts://` URI scheme
+- No S3/GCS credentials required
+- Artifact viewing in UI works seamlessly
+
+#### Story 2: Operator Deploying KFP Without External Object Storage
+
+As an operator for a Kubeflow distribution, I want to deploy KFP with filesystem storage so that I don't need to productize and support a separate object storage system.
+
+**Acceptance Criteria:**
+
+- Single configuration option to enable filesystem storage
+- Artifact handling is part of KFP (no separate object storage component)
+- Storage automatically provisioned via PVCs
+- Backup/restore follows standard Kubernetes PVC procedures
+
+#### Story 3: Operator Configuring Storage Class and Size
+
+As an operator, I want to configure KFP to use a specific StorageClass and PVC size instead of defaults, so that I can match storage performance and capacity to my workload requirements.
+
+**Acceptance Criteria:**
+
+- Can specify `StorageClass` in KFP configuration
+- Can set PVC size limits (global configuration)
+- Storage quotas enforced via Kubernetes `ResourceQuotas`
+- Clear error messages when storage limits are reached
+- Can choose between RWO and RWX access modes based on needs
```

Comment:
```suggestion
```

That is already covered by the default seaweedfs with zero effort. You can already decide which storageclass is used for Seaweedfs. It already satisfies
- Can specify `StorageClass` in KFP configuration
- Can set PVC size limits (global configuration)
- Clear error messages when storage limits are reached
- Can choose between RWO and RWX access modes based on needs

So i recommend to remove this story

The only thing remaining is - Storage quotas enforced via Kubernetes `ResourceQuotas`. We can set it per bucket already, but i am not sure about per folder. So it is partially solved and fully solved if you create on bucket per namespace. This item can be moved to another story.